server/migrations/1613615463607_reassign-participant-regions.js

```diff
@@ -0,0 +1,71 @@
+/* eslint-disable no-restricted-syntax, no-await-in-loop, max-len */
+const { readFileSync } = require('fs');
+const { join } = require('path');
+const readXlsxFile = require('node-xlsx').default;
+const { dbClient } = require('../db/db.js');
+const { collections } = require('../db/schema.js');
+
+const columnNames = ['ClientID', 'EOI - FHA', 'EOI - IHA', 'EOI - NHA', 'EOI - VCHA', 'EOI - VIHA'];
+const columnValues = ['Fraser', 'Interior', 'Northern', 'Vancouver Coastal', 'Vancouver Island'];
+
+// Program area provided a spreadsheet with participants' preferred health regions as columns
+// These columns contain textual quasi-booleans to denote whether the respective region is preferred
+// Booleans look like "YES" for truthy and "NO" or "NULL" for falsy (all case-insensitive)
+const textualBoolean = (v) => {
+  if (typeof v !== 'string') throw Error(`Invalid boolean: ${v} (type: ${typeof v})`);
+  switch (v.trim().toUpperCase()) {
+    case 'YES':
+      return true;
+    case 'NULL':
+    case 'NO':
+      return false;
+    default:
+      throw Error(`Invalid boolean string: ${v}`); // Better safe than sorry
+  }
+};
+
+exports.up = async () => {
+  const file = readFileSync(join(__dirname, 'assets', 'reassign-participant-regions.xlsx'));
+  const sheet = readXlsxFile.parse(file, { raw: true })?.[0]?.data; // Only need the first sheet
+  const [headers, ...data] = sheet; // First line is headers
+
+  if (!headers.every((e, i) => columnNames[i] === e)) throw Error('Unexpected headers');
+
+  const changes = data.map((row) => ({
+    id: row?.[0],
+    preferredLocation: row
+      .slice(1) // Ignore Maximus ID column
+      .map((v, i) => (textualBoolean(v) ? columnValues[i] : null)) // Map from textual boolean to HA string value
```

Comment:
So this is not something we should change, given our timeline on this fix and the fact that it's a migration (meaning we are not depending on this code elsewhere and don't need to update/maintain it)

But I do think this is a good teaching/learning thing to point out - this is a really brittle design (again caveat, I know this was written in a hurry).

We should just be storing these values in a map to begin with, since that's fit for purpose, instead of relying on the order happening to be correct and the lengths of the arrays happening to match. This is the kind of thing that becomes very prone to breaking in code that gets used elsewhere in the app and maintained over time. Once you have things in a map, it's easy enough to just pull the keyset out if you only want that or vice versa, and now the mapping is a lot more robust and resistant to missed assumptions.

Reply:
lol as soon as I write this I see we do check the order on line 32

I still really don't like assuming no one is going to touch your arrays though! Totally doesn't matter here but 👻